src/core/reactor_backend.cc

```diff
@@ -1484,10 +1490,96 @@ class reactor_backend_uring final : public reactor_backend {
+                delete this;
+            }
+            void set_exception(std::exception_ptr eptr) noexcept final {
+                _result.set_exception(eptr);
+                delete this;
+            }
+            future<std::tuple<pollable_fd, socket_address>> get_future() {
+                return _result.get_future();
+            }
+            ::sockaddr* posix_sockaddr() {
+                return &_sa.as_posix_sockaddr();
+            }
+            socklen_t* socklen_ptr() {
+                return &_sa.addr_length;
+            }
+        };
+        auto desc = std::make_unique<accept_completion>(listenfd);
+        auto req = internal::io_request::make_accept(listenfd.fd.get(), desc->posix_sockaddr(), desc->socklen_ptr(), SOCK_CLOEXEC);
+        auto fut = desc->get_future();
+        _r._io_sink.submit(desc.release(), std::move(req));
+        return fut.handle_exception_type([&listenfd] (const std::system_error &e) {
+             if (e.code() == std::errc::invalid_argument) {
+                 // The chances are that we shutting down the connection.
+                 listenfd.maybe_no_more_recv();
+             }
+             return make_exception_future<std::tuple<pollable_fd, socket_address>>(e);
+        });
     }
     virtual future<> connect(pollable_fd_state& fd, socket_address& sa) override {
-        return _r.do_connect(fd, sa);
+        class connect_completion final : public io_completion {
+            socket_address _sa;
+            promise<> _result;
+        public:
+            void complete(size_t fd) noexcept final {
+                _result.set_value();
+                delete this;
+            }
+            void set_exception(std::exception_ptr eptr) noexcept final {
+                _result.set_exception(eptr);
+                delete this;
+            }
+            future<> get_future() {
+                return _result.get_future();
+            }
+            ::sockaddr* posix_sockaddr() {
+                return &_sa.as_posix_sockaddr();
+            }
+            socklen_t socklen() const {
+                return _sa.addr_length;
+            }
+        };
+        auto desc = std::make_unique<connect_completion>();
+        auto req = internal::io_request::make_connect(fd.fd.get(), desc->posix_sockaddr(), desc->socklen());
+        auto fut = desc->get_future();
+        _r._io_sink.submit(desc.release(), std::move(req));
+        return fut.handle_exception_type([] (const std::system_error &e) {
+            if (e.code() == std::errc::operation_in_progress) {
+                return make_ready_future();
+            }
+            throw;
+        }).then([&fd, this] {
+            return writeable(fd);
+        }).then([&fd] {
+            auto err = fd.fd.getsockopt<int>(SOL_SOCKET, SO_ERROR);
+            if (err != 0) {
+                throw std::system_error(err, std::system_category());
+            }
+            return make_ready_future();
+        });
```

Comment:
Please make sure to test both successful and unsuccessful connect() (e.g. with seawreck).

Reply:
sure. tested with `apps/seawreck/seawreck --smp 1 -s 192.168.2.8:10000 --reactor-backend io_uring`. the behaviors of successful and unsuccessful connect() are the same with those tested with `--reactor-backend linux-aio`.